automation_tools/satellite6/log.py

```diff
@@ -0,0 +1,106 @@
+    [root@sathost.redhat.com] Executing task 'get_line_count'
+    [root@sathost.redhat.com] Executing task 'fetch_appended_log_lines'
+
+    """
+
+    def __init__(self, sat_host):
+        """Initializes context manager with Satellite hostname
+
+        :param sat_host: str
+        """
+        self.host = sat_host
+        self.log_state = {
+            '/var/log/foreman-installer/satellite.log': 0,
+            '/var/log/foreman/production.log': 0,
+            '/var/log/foreman-proxy/proxy.log': 0,
+            '/var/log/candlepin/candlepin.log': 0,
+            '/var/log/messages': 0,
+            '/var/log/mongodb/mongodb.log': 0,
+            '/var/log/tomcat/catalina.out': 0
+        }
+
+    def __enter__(self):
+        """
+        Fetch current line count for Satellite log files
+        :return: LogAnalyzer
+        """
+        self._update_log_files_state()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Analyzes log files checking if some error occurred since last
+        log_state update
+        """
+        print('#### Analyzing logs of host %s' % self.host)
+        self._update_log_files_state()
+
+        def fetch_appended_log_lines():
+            for log_file, lines_appended in self.log_state.items():
+                if lines_appended > 0:
+                    content = run(
+                        'tail -n {} {}'.format(lines_appended, log_file),
+                        quiet=True
+                    )
+                    analyze(log_file, content)
+
+        execute(fetch_appended_log_lines, host=self.host)
+
+    def _update_log_files_state(self):
+        """Update log_dct with adding delta from current number of lines of
+        each item and the last provided by dct.
+
+        So this method can be used to check how many lines were appended on
+        a file during some processes and used to tail them. If log_state is
+        None a new dict is created with initial values been 0 (zero).
+
+        :param host: str with satellite hostname
+        :param log_dct: dict to be updated
+        """
+
+        def get_line_count():
+            for log_file, old_value in self.log_state.items():
+                try:
+                    current_value = int(
+                        run(
+                            'wc -l < {}'.format(log_file),
+                            quiet=True
+                        )
+                    )
+                except ValueError:
+                    self.log_state[log_file] = 0
+                else:
+                    self.log_state[log_file] = current_value - old_value
+
+        execute(get_line_count, host=self.host)
+
+
+def analyze(log_file, content):
+    """Analyzes appended content from a log file. For now it is only
+    checking for ERROR status on regular log files.
+    """
+    _print_wrapper('### Analyzing %s:' % log_file)
+    errors = sum(map(lambda _: 1, re.finditer('ERROR', content)))
```

Comment:
@renzon :+1:  for using map, RE and sum for finding the number of total Errors. What a convenient way to use them . You are the real python champ.